[core] Vendor aiohttp and aiosignal for Ray.

rynewang opened the pull request:
This PR moves `aiohttp`, `aiohttp-cors` and `aiosignal` from `Ray[default]` dependencies to vendored `third_party` libraries.

Previously we already vendored `aiohttp` for runtime env agent. It worked well but was unnecessarily restricted for runtime env agent to use, and the dashboard http server still depends on it via `ray[default]`. This PR removes the runtime env agent specific vendor dir `ray/_private/runtime_env/agent/thirdparty_files` and moves `aiohttp` into the Ray shared `ray/thirdparty_files`.

Also vendors `aiosignal`.

FIxes #40339.

pcmoritz commented:
Thanks for putting this together -- we need to do it in a way that makes these dependencies only visible in the dashboard and runtime environment agent processes (i.e. it shouldn't be visible to workers or any other processes that the user can put code into -- they should be able to use whichever version of these packages they want) :)

rynewang commented:
@pcmoritz just to confirm, we will do things like this:

- declared deps: like `frozenlist` and `requests`, we use same version as users, as requested in pip pkg definition
- vendored for Ray: ["psutil", "setproctitle==1.2.2", "colorama"], if user use them in Worker we choose our vendored version
- vendored for Dash: aiohttp, aiosignal, if user use them in Worker they use their own; only Dash and Runtime Env Agent use it

pcmoritz commented:
Yes that is exactly right. At a high level, we should reduce the amount of vendored packages in worker code to the absolute minimum since it can interfere with user dependencies / code (whereas dashboard doesn't).

In addition to what you say, we should also vendor frozenlist in the dashboard and remove it from the dependencies since it is only used in the dashboard :)

Thanks for doing all this work :)

pcmoritz reviewed (approved):
This looks great, thanks for doing it! There are a few test failures you need to fix before we can merge it :)

aslonnie commented on python/setup.py:
```diff
@@ -535,18 +536,22 @@ def build(build_python, build_java, build_cpp):
-    runtime_env_agent_pip_packages = ["aiohttp"]
+    # internal dependenceis
+    internal_pip_packages = [
+        "aiohttp >= 3.7",  # for runtime env agent and the dashboard http server
+        "aiohttp_cors",  # for dashboard
+        "aiosignal",  # for dashboard
```
you need to pin a version here.

aslonnie reviewed (commented):
for vendor'ed packages, it might be better to just copy and check in the files, rather than trying to do the pip install into a directory thing.

jjyao commented on dashboard/http_server_head.py:
```diff
@@ -16,11 +16,8 @@
-# installation must be included in this file. This allows us to determine if
-# the agent has the necessary dependencies to be started.
-from ray.dashboard.optional_deps import aiohttp, hdrs
+from ray._private.internal_third_party import aiohttp  # noqa: F401
```
We already have `python/ray/_private/thirdparty/xxx`. What's the difference between that folder and this folder?

jjyao commented:
Seems pip has a tool https://pypi.org/project/vendoring/ that it uses to vendor its dependencies. Something we can see if we can leverage.

rynewang commented:
So now we have these 4 kinds of third party codes:


|                          | where                          | visible to users        | used in                                                  | pkgs                                  |
|--------------------------|--------------------------------|-------------------------|----------------------------------------------------------|---------------------------------------|
| declared dependencies    | $PYTHONPATH                    | yes                     | everywhere                                               | a lot, e.g. jsonschema and protobuf   |
| in-tree vendors          | ray/_private/thirdparty        | no?                     | everywhere                                               | dacite, pathspec, pynvml, tabulate    |
| "public vendors"         | ray/thirdparty_files           | yes after  `import ray` | everywhere                                               | psutil, setproctitle==1.2.2, colorama |
| (NEW!) "private vendors" | ray/_private/third_party_files | no                      | only Ray internal processes (Dashboard, RuntimeEnvAgnet) | aiohttp, aiohttp_cors, aiosignal      |


I agree this is a little bit scattered. How do we reconcile? @pcmoritz @aslonnie @jjyao

rynewang commented:
It does not look like we can effortlessly add big packages like aiohttp to the in-tree vendors. We need to fix all `import aiohttp.a.b` in the package itself to relative imports `import .aiohttp.a.b` notice the `.` before `aiohttp`. This applies transitively to all its dependencies. And we will bear the burden to do this every time we update the deps.

rynewang commented:
The `vendoring` readme states:

> This tool has no stability promises -- it has only one intended user: pip. There may be unannounced changes to this codebase at any time, as long as the intended user (i.e. the pip project) is prepared for those changes.
> 
> As a general rule of thumb, if the project is going to be a PyPI package, it should not use this tool.

so I guess we need something else.

rkooo567 commented:
Not super familiar with bazel, but can't we actually just vendor all deps to thirdparty_files using the same way as we vendor cpp deps? https://github.com/ray-project/ray/blob/6948c204055def067af598cdbd37d6399b8731ae/bazel/ray_deps_setup.bzl#L135.

rynewang commented:
Updated. Changes:

- Pinned all versions according to requirements_compiled.txt
- Discovered that Serve actually uses aiohttp. Added back dependency to `ray[serve]` (Removed from `ray[default]`).
- Restricted access to the internal deps to Dashboard, Runtime Env Agent, Dashboard tests, and release script (`release/ray_release/command_runner/_prometheus_metrics.py`).

can-anyscale reviewed (commented):
not sure if you're still working on it, comment to remove from my review queue, feel free to add me again if needed

aslonnie reviewed (commented):
(drop review)
